### PR TITLE
【evm】Support using trie for storing native token balances

### PR DESCRIPTION
**Pull request by 516108736**

### CODE REVIEW

File: core/state/iterator.go [hunk 1]
```diff
@@ -105,6 +107,7 @@ func (it *NodeIterator) step() error {
 	}
 	// Otherwise we've reached an account node, initiate data iteration
 	var account Account
+	account.TokenBalances, _ = types.NewTokenBalances(nil, it.state.db.TrieDB())
```
Comment (516108736) [inline core/state/iterator.go]: only used in test

File: core/state/journal.go [hunk 2]
```diff
@@ -154,7 +154,7 @@ func (ch suicideChange) revert(s *StateDB) {
 	obj := s.getStateObject(*ch.account)
 	if obj != nil {
 		obj.suicided = ch.prev
-		obj.SetBalances(ch.prevbalance)
+		obj.setBalances(ch.prevbalance)
```
Comment (516108736) [inline core/state/journal.go]: bug fix

File: core/state/sync.go [hunk 3]
```diff
@@ -19,16 +19,21 @@ package state
...
 	"github.com/ethereum/go-ethereum/trie"
 )
 
 // NewStateSync create a new state trie download scheduler.
-func NewStateSync(root common.Hash, database trie.DatabaseReader) *trie.Sync {
+func NewStateSync(root common.Hash, database ethdb.Database) *trie.Sync {
 	var syncer *trie.Sync
 	callback := func(leaf []byte, parent common.Hash) error {
 		var obj Account
+		obj.TokenBalances, _ = types.NewTokenBalances(nil, trie.NewDatabase(database))
```
Comment (516108736) [inline core/state/sync.go]: only for test

**Comment by 516108736** [comment]
https://github.com/QuarkChain/pyquarkchain/pull/785/files

File: core/types/token.go [hunk 4]
```diff
@@ -100,104 +106,175 @@ func NewTokenBalances(data []byte) (*TokenBalances, error) {
...
+			return true
+		}
+
+		if t.tokenTrie != nil && len(t.balances) == 0 {
+			return true
+		}
+		return false
+	}
+	if !readyBeforeSer() {
+		return nil, errors.New("bug here")
```
Comment (pingke) [inline core/types/token.go]: will error when t.tokenTrie != nil and setValue call

File: core/types/token.go [hunk 5]
```diff
@@ -100,104 +106,175 @@ func NewTokenBalances(data []byte) (*TokenBalances, error) {
...
-	default:
-		return nil, fmt.Errorf("Unknown enum byte in token_balances")
-
+		list = append(list, &TokenBalancePair{
+			TokenID: k,
+			Balance: v,
+		})
+	}
+	sort.Slice(list, func(i, j int) bool { return list[i].TokenID < (list[j].TokenID) })
+	rlpData, err := rlp.EncodeToBytes(list)
```
Comment (pingke) [inline core/types/token.go]: make w as writer

**Pull request merged**